core/bin/external_node/src/version_sync_task.rs

```diff
@@ -0,0 +1,132 @@
+use std::cmp::Ordering;
+
+use anyhow::Context;
+use zksync_basic_types::{L1BatchNumber, MiniblockNumber};
+use zksync_dal::ConnectionPool;
+use zksync_types::ProtocolVersionId;
+use zksync_web3_decl::{
+    jsonrpsee::http_client::HttpClient,
+    namespaces::{EnNamespaceClient, ZksNamespaceClient},
+};
+
+pub async fn get_l1_batch_remote_protocol_version(
+    main_node_client: &HttpClient,
+    l1_batch_number: L1BatchNumber,
+) -> anyhow::Result<Option<ProtocolVersionId>> {
+    let Some((miniblock, _)) = main_node_client
+        .get_miniblock_range(l1_batch_number)
+        .await?
+    else {
+        return Ok(None);
+    };
+    let sync_block = main_node_client
+        .sync_l2_block(MiniblockNumber(miniblock.as_u32()), false)
+        .await?;
+    Ok(sync_block.map(|b| b.protocol_version))
+}
+
+// Synchronizes protocol version in `l1_batches` and `miniblocks` tables between EN and main node.
+pub async fn sync_versions(
+    connection_pool: ConnectionPool,
+    main_node_client: HttpClient,
+) -> anyhow::Result<()> {
+    tracing::info!("Starting syncing protocol version of blocks");
+
+    let mut connection = connection_pool.access_storage().await?;
+
+    // Load the first local batch number with version 22.
+    let Some(local_first_v22_l1_batch) = connection
+        .blocks_dal()
+        .get_first_l1_batch_number_for_version(ProtocolVersionId::Version22)
+        .await?
+    else {
+        return Ok(());
+    };
+    tracing::info!("First local v22 batch is #{local_first_v22_l1_batch}");
+
+    // Find the first remote batch with version 22, assuming it's less then or equal than local one.
+    // Uses binary search.
+
+    let mut left_bound = L1BatchNumber(0);
+    let mut right_bound = local_first_v22_l1_batch;
+
+    let right_bound_remote_version =
+        get_l1_batch_remote_protocol_version(&main_node_client, right_bound).await?;
```

Comment:
I think, that method in zks namespace: 
`getProtocolUpgradeBlock` is pretty useful, we can implement it for a better system overview.

Reply:
Discussed in DM and decided to leave current approach with binary search